packages/js/core/src/types/WrapError.ts

```diff
@@ -0,0 +1,212 @@
+import { CleanResolutionStep } from "../algorithms";
+
+/**
+Wrap error codes provide additional context to WrapErrors.
+
+Error code naming convention (approximate):
+    type of handler
+    type of functionality
+    piece of functionality
+    ==> handler_typeFn_pieceFn
+
+Error code map:
+  0-24 -> Client
+  25-49 -> URI resolution
+  50-74 -> Wrapper invocation & sub-invocation
+  75-255 -> Unallocated
+ */
+export enum WrapErrorCode {
+  CLIENT_LOAD_WRAPPER,
+  CLIENT_GET_FILE,
+  CLIENT_GET_IMPLEMENTATIONS,
+  CLIENT_VALIDATE_RESOLUTION,
+  CLIENT_VALIDATE_ABI,
+  CLIENT_VALIDATE_RECURSIVE,
+  URI_RESOLUTION = 25,
+  URI_RESOLVER,
+  URI_NOT_FOUND,
+  WRAPPER_INVOKE_ABORTED = 50,
+  WRAPPER_SUBINVOKE_ABORTED,
+  WRAPPER_INVOKE_FAIL,
+  WRAPPER_READ_FAIL,
+  WRAPPER_STATE_ERROR,
```

Comment:
I would rather call this something more universal / catch-all, that could be used in all types of wrappers. Basically what it is being used for now is to catch all internal errors of the wasm wrapper that should never happen. Effectively it's for bad internal states that shouldn't exist, but could possibly if the internal mechanisms within the wrapper fail in some way.

So, I think we should rename this to `WRAPPER_INTERNAL_ERROR`, so that it can be used to represent any unforeseen errors within the wrapper's internals. These errors shouldn't happen if the wrapper was functioning as intended, but could given extreme circumstances. This could be used for `PluginWrapper`, `WasmWrapper`, and even an `HttpWrapper` if we build that in the future. For example `HttpWrapper` could throw an internal error if there are no networking capabilities available on the device. These internal errors can be made more "colorful" by doing the `WrapError` sub-classing like you mentioned above.

Idk, just a thought, curious what you think.

Reply:
Very good thoughts. I renamed `WRAPPER_STATE_ERROR` to `WRAPPER_INTERNAL_ERROR`.